File: apps/mac/Sources/Design/Skeleton.swift
// SPDX-License-Identifier: LicenseRef-tokenstat-source-available
//
// Source-available for review, NOT open source. See LICENSE: no rights to
// redistribute, publish, or ship a build are granted. Read it, study it, run
// your own build of it.
// "tokenstat" is a trademark of pueev OU. See TRADEMARK.md.

import SwiftUI

/// Placeholders shaped like the thing that is coming.
///
/// Flow for a cold launch:
/// 1. App splash (logo) while the host starts — see `LaunchState`.
/// 2. **These** wireframes, with a light pulse so the wait does not feel frozen.
/// 3. Real content replaces each skeleton with `.smoothIn`.
///
/// Never a spinner in an empty pane, and never a blurred wireframe under a
/// brand mark on the same layer.
enum Skeleton {
    /// One grey bar standing in for a line of text or a number.
    ///
    /// `width` of nil fills the space, which is what a title or a row wants.
    /// A soft opacity pulse (when motion is allowed) keeps the layout alive
    /// without a shimmer that steals attention from the data about to land.
    struct Bar: View {
        var width: CGFloat?
        var height: CGFloat = 12
        /// Phase offset so neighbouring bars do not pulse in lockstep.
        var phase: Double = 0

        @Environment(\.accessibilityReduceMotion) private var reduceMotion
        @State private var bright = false

        var body: some View {
            RoundedRectangle(cornerRadius: 4)
                .fill(Theme.border)
                .frame(width: width, height: height)
                .frame(maxWidth: width == nil ? .infinity : nil, alignment: .leading)
                .opacity(reduceMotion ? 1 : (bright ? 1 : 0.52))
                .animation(
                    reduceMotion
                        ? nil
                        : .easeInOut(duration: 0.95)
                            .repeatForever(autoreverses: true)
                            .delay(phase),
                    value: bright
                )
                .onAppear {
                    guard !reduceMotion else { return }
                    bright = true
                }
        }
    }

    /// A stack of rows, each a label bar and a value bar, like a list.
    ///
    /// Widths vary down the stack so it reads as text rather than as a bar
    /// chart. Deterministic, not random: a placeholder that reshuffles on every
    /// redraw is a distraction of its own.
    struct Rows: View {
        var count: Int = 5
        var showsValue = true

        var body: some View {
            VStack(spacing: Theme.Space.s) {
                ForEach(0..<count, id: \.self) { index in
                    HStack(spacing: Theme.Space.s) {
                        Bar(
                            width: Self.labelWidths[index % Self.labelWidths.count],
                            phase: Double(index) * 0.08
                        )
                        Spacer(minLength: Theme.Space.s)
                        if showsValue {
                            Bar(width: 52, phase: Double(index) * 0.08 + 0.04)
                        }
                    }
                }
            }
        }

        private static let labelWidths: [CGFloat] = [128, 96, 152, 84, 116, 104]
    }

    /// A card-shaped placeholder: heading, subheading, and some rows.
    struct CardPlaceholder: View {
        var rows: Int = 3

        var body: some View {
            VStack(alignment: .leading, spacing: Theme.Space.m) {
                VStack(alignment: .leading, spacing: 4) {
                    Bar(width: 116, height: 11, phase: 0)
                    Bar(width: 168, height: 9, phase: 0.06)
                }
                Rows(count: rows)
            }
            .padding(Theme.cardPadding)
            .frame(maxWidth: .infinity, alignment: .topLeading)
            .background(Theme.panel, in: RoundedRectangle(cornerRadius: Theme.cardRadius))
            .overlay(
                RoundedRectangle(cornerRadius: Theme.cardRadius)
                    .strokeBorder(Theme.border, lineWidth: 1)
            )
        }
    }
}

/// Marks a placeholder subtree as non-interactive while data is loading.
///
/// Wireframes stay sharp (pulse lives on `Skeleton.Bar`). Real content replaces
/// them with `.smoothIn`.
struct Warming: ViewModifier {
    let active: Bool

    func body(content: Content) -> some View {
        content.allowsHitTesting(!active)
    }
}

extension View {
    /// Mark this subtree as a placeholder for data that has not arrived.
    func warming(_ active: Bool) -> some View {
        modifier(Warming(active: active))
    }
}

extension AnyTransition {
    /// The arrival of loaded content: a short fade with a small rise.
    /// Collapses to a plain fade when Reduce Motion is on.
    static func smoothIn(reduceMotion: Bool) -> AnyTransition {
        reduceMotion ? .opacity : .opacity.combined(with: .offset(y: 4))
    }
}

File: apps/android/app/src/main/java/ai/tokenstat/tokenstat/ui/components/Skeleton.kt
// SPDX-License-Identifier: LicenseRef-tokenstat-source-available
package ai.tokenstat.tokenstat.ui.components

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.StartOffset
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import ai.tokenstat.tokenstat.ui.theme.LocalTsColors
import ai.tokenstat.tokenstat.ui.theme.Space
import ai.tokenstat.tokenstat.ui.theme.TsMotion
import ai.tokenstat.tokenstat.ui.theme.cardPadding
import ai.tokenstat.tokenstat.ui.theme.cardRadius
import ai.tokenstat.tokenstat.ui.theme.rememberReduceMotion

private val barShape = RoundedCornerShape(4.dp)

/// One grey bar standing in for a line of text or a number.
///
/// [width] of null fills the space, which is what a title or a row wants. A
/// soft opacity pulse (when motion is allowed) keeps the layout alive without
/// a shimmer that steals attention from the data about to land. [phaseMillis]
/// offsets neighbouring bars so they do not pulse in lockstep.
@Composable
fun SkeletonBar(width: Dp? = null, height: Dp = 12.dp, phaseMillis: Int = 0) {
    val colors = LocalTsColors.current
    val reduceMotion = rememberReduceMotion()
    val transition = rememberInfiniteTransition(label = "skeleton")
    val alpha by transition.animateFloat(
        initialValue = 0.52f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(TsMotion.pulseMillis, easing = TsMotion.easeInOut),
            repeatMode = RepeatMode.Reverse,
            initialStartOffset = StartOffset(phaseMillis),
        ),
        label = "skeletonPulse",
    )
    Box(
        Modifier
            .then(if (width != null) Modifier.width(width) else Modifier.fillMaxWidth())
            .height(height)
            .clip(barShape)
            .background(colors.border)
            .then(if (reduceMotion) Modifier else Modifier.alpha(alpha)),
    )
}

/// A stack of rows, each a label bar and a value bar, like a list. Widths vary
/// down the stack so it reads as text rather than as a bar chart.
/// Deterministic, not random: a placeholder that reshuffles on every redraw is
/// a distraction of its own.
@Composable
fun SkeletonRows(count: Int = 5, showsValue: Boolean = true) {
    val labelWidths = listOf(128.dp, 96.dp, 152.dp, 84.dp, 116.dp, 104.dp)
    Column(verticalArrangement = Arrangement.spacedBy(Space.s)) {
        repeat(count) { index ->
            Row(verticalAlignment = Alignment.CenterVertically) {
                SkeletonBar(
                    width = labelWidths[index % labelWidths.size],
                    phaseMillis = index * 80,
                )
                Spacer(Modifier.weight(1f))
                if (showsValue) SkeletonBar(width = 52.dp, phaseMillis = index * 80 + 40)
            }
        }
    }
}

/// A card-shaped placeholder: heading, subheading, and some rows.
@Composable
fun SkeletonCard(rows: Int = 3) {
    val colors = LocalTsColors.current
    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(cardRadius))
            .background(colors.panel)
            .tsCardBorder()
            .padding(cardPadding),
        verticalArrangement = Arrangement.spacedBy(Space.m),
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            SkeletonBar(width = 116.dp, height = 11.dp, phaseMillis = 0)
            SkeletonBar(width = 168.dp, height = 9.dp, phaseMillis = 60)
        }
        SkeletonRows(count = rows)
    }
}
